finality-aleph/src/testing/data_io.rs

```diff
@@ -0,0 +1,385 @@
+use futures::{
+    channel::{
+        mpsc::{self, UnboundedReceiver, UnboundedSender},
+        oneshot,
+    },
+    StreamExt,
+};
+use sc_block_builder::BlockBuilderProvider;
+use sp_api::BlockId;
+use sp_api::NumberFor;
+use sp_consensus::BlockOrigin;
+use sp_runtime::traits::Block as BlockT;
+use sp_runtime::Digest;
+use std::{default::Default, future::Future, sync::Arc, time::Duration};
+use substrate_test_runtime_client::{
+    runtime::Block, Backend, ClientBlockImportExt, DefaultTestClientBuilderExt, TestClient,
+    TestClientBuilder, TestClientBuilderExt,
+};
+use tokio::sync::Mutex;
+
+#[derive(Clone)]
+struct TestBlockRequester<B: BlockT> {
+    blocks: mpsc::UnboundedSender<AlephDataFor<B>>,
+    justifications: mpsc::UnboundedSender<AlephDataFor<B>>,
+}
+
+impl<B: BlockT> TestBlockRequester<B> {
+    fn new() -> (
+        Self,
+        UnboundedReceiver<AlephDataFor<B>>,
+        UnboundedReceiver<AlephDataFor<B>>,
+    ) {
+        let (blocks_tx, blocks_rx) = mpsc::unbounded();
+        let (justifications_tx, justifications_rx) = mpsc::unbounded();
+        (
+            TestBlockRequester {
+                blocks: blocks_tx,
+                justifications: justifications_tx,
+            },
+            blocks_rx,
+            justifications_rx,
+        )
+    }
+}
+
+impl<B: BlockT> RequestBlocks<B> for TestBlockRequester<B> {
+    fn request_justification(&self, hash: &B::Hash, number: NumberFor<B>) {
+        self.justifications
+            .unbounded_send(AlephData {
+                hash: *hash,
+                number,
+            })
+            .unwrap();
+    }
+
+    fn request_stale_block(&self, hash: B::Hash, number: NumberFor<B>) {
+        self.blocks
+            .unbounded_send(AlephData { hash, number })
+            .unwrap();
+    }
+}
+
+type TestData = Vec<AlephDataFor<Block>>;
+
+impl AlephNetworkMessage<Block> for TestData {
+    fn included_blocks(&self) -> Vec<AlephDataFor<Block>> {
+        self.clone()
+    }
+}
+
+#[derive(Clone)]
+struct TestNetwork {
+    sender_tx: UnboundedSender<(TestData, Recipient)>,
+    next_rx: Arc<Mutex<UnboundedReceiver<TestData>>>,
+}
+
+impl TestNetwork {
+    fn new() -> (
+        Self,
+        UnboundedReceiver<(TestData, Recipient)>,
+        UnboundedSender<TestData>,
+    ) {
+        let (sender_tx, sender_rx) = mpsc::unbounded();
+        let (next_tx, next_rx) = mpsc::unbounded();
+        (
+            TestNetwork {
+                sender_tx,
+                next_rx: Arc::new(Mutex::new(next_rx)),
+            },
+            sender_rx,
+            next_tx,
+        )
+    }
+}
+
+#[async_trait]
+impl ComponentNetwork<TestData> for TestNetwork {
+    type S = UnboundedSender<(TestData, Recipient)>;
+    type R = UnboundedReceiver<TestData>;
+
+    fn sender(&self) -> &Self::S {
+        &self.sender_tx
+    }
+
+    fn receiver(&self) -> Arc<Mutex<Self::R>> {
+        self.next_rx.clone()
+    }
+}
+
+struct DataStoreChannels {
```

Comment:
Suggestion – rename this to `DataStoreIO`, remove the `network` field and add methods like `block_request()`, `send_data(data)`, and `exit()`. This should make the code below more readable (you are always calling methods on the `io` object when you want to interact with the data store), remove the `Box<dyn ...>` (just return `impl DataNetwork<...>` from the function), and remove the unpacking boilerplate from the test cases.

Reply:
On the other hand maybe `TestClient` should be included here and `import_blocks` could be a method too. There would need to be some block building methods too, but I'm pretty sure they should be functions anyway (see some comments below), so that is only an advantage.

Reply:
Done. I moved all this into one struct and named it `TestHandler`. It has all necessary methods to carry out the tests.